_posts/2018-12-06.Rust-1.31.md

```diff
@@ -0,0 +1,338 @@
+This release adds a new way to qualify a function: `const fn`. It looks like
+this:
+
+```rust
+const fn foo(x: i32) -> i32 {
+    x + 1
+}
+```
+
+A `const fn` can be called like a regular function, but it can also be used
+in any constant context. When it is, it is evaluated at compile time, rather
+than at run time. As an example:
+
+```rust
+const SIX: i32 = foo(5);
+```
+
+This will execute `foo` at compile time, and set `SIX` to `6`.
+
+`const fn`s cannot do everything that normal `fn`s can do; they must
+have deterministic output. This is important for soundness reasons.
+Currently, `const fn`s can do a minimal subset of operations. Here's
+some examples of what you can do:
+
+* Arithmetic and comparison operators on integers
+* All boolean operators except for `&&` and `||`
+* Constructing arrays, structs, enums, and tuples
+* Calls to other `const fn`s
+* Index expressions on arrays and slices
+* Field accesses on structs and tuples
+* Reading from constants (but not statics, not even taking a reference to a static)
+* `&` and `*` of references
+* Casts, except for raw pointer to integer casts
+
+We'll be slowly growing the abilities of `const fn`, but we've decided that
+this is enough useful stuff to start shipping the feature itself.
+
+For full details, please see [the
+reference](https://doc.rust-lang.org/reference/items/functions.html#const-functions).
+
+### Tool lints
+
+In [Rust 1.30](https://blog.rust-lang.org/2018/10/25/Rust-1.30.0.html), we
+stabilized "tool attributes", like `#[rustfmt::skip]`. In Rust 1.31, we're
+stabilizing something similar: "tool lints," like
+`#[allow(clippy::something)]` These give a namespace to lints, so that it's
+more clear which tool they're coming from.
+
+If you previously used Clippy's lints, you can migrate like this:
+
+```rust
+// old
+#![allow(bool_comparison)]
+
+// new
+#![allow(clippy::bool_comparison)]
+```
+
+You'll get warnings that can help you update to the new style.
+
+### Library stabilizations
+
+A bunch of `From` implementations have been added:
+
+* `u8` now implements `From<NonZeroU8>`, and likewise for the other numeric types and their `NonZero` equivalents
+* `Option<&T>` implements `From<&Option<T>>`, and likewise for `&mut`
+
+Additionally, these functions have been stabilized:
+
+* [`slice::align_to`](https://doc.rust-lang.org/std/primitive.slice.html#method.align_to) and its mutable counterpart
+* [`slice::chunks_exact`](https://doc.rust-lang.org/std/primitive.slice.html#method.chunks_exact),
+  as well as its mutable and `r` counterparts (like
+  [`slice::rchunks_exact_mut`](https://doc.rust-lang.org/std/primitive.slice.html#method.rchunks_mut)) in all combinations
+
+See the [detailed release notes][notes] for more.
+
+### Cargo features
+
+Cargo will now download packages in parallel using HTTP/2. Additionally,
+now that `extern crate` is gone, you can't `extern crate foo as bar;` anymore
```

Comment:
What do we mean by "is gone"?

Reply:
I think a better wording would be "Now that extern crate is no longer completely idiomatic" or something?

Reply:
"Now that `extern crate` is not usually required, it would be jarring to have to use `extern crate foo as bar;` to import crate with a different name."?